ts: Properly read columnar format

mrtracy opened the pull request:
This commit allows the time series query system to properly handle
on-disk data stored in the columnar format. It does so in a way that is
completely backwards compatible with the row-based format; a queried
time span can actually handle a mixture of row-and-column formatted data
blocks and produce a correct response.

This does not yet include the actual writing of columnar data to disk -
all data being written is currently still in row format. As such, tests
have not yet been added verifying the columnar format through the entire
pipeline. However, unit tests for the downsampling logic have been
greatly expanded, as this was the portion of the code most greatly
affected by this change.

Release note: none.

Commit:
ts: Properly read columnar format

This commit allows the time series query system to properly handle
on-disk data stored in the columnar format. It does so in a way that is
completely backwards compatible with the row-based format; a queried
time span can actually handle a mixture of row-and-column formatted data
blocks and produce a correct response.

This does not yet include the actual writing of columnar data to disk -
all data being written is currently still in row format. As such, tests
have not yet been added verifying the columnar format through the entire
pipeline. However, unit tests for the downsampling logic have been
greatly expanded, as this was the portion of the code most greatly
affected by this change.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26446)
<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/roachpb/internal.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJW4Pp1S9al83y-fYj:-LEJW4Pp1S9al83y-fYk:b1ers78) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/roachpb/internal.go#L61)):*
> ```Go
> // in columnar format.
> func (data *InternalTimeSeriesData) IsRollup() bool {
> 	return len(data.Offset) > 0
> ```

Are we going to be storing all samples in columnar format when you're done, or only rollups?

---

*[pkg/roachpb/internal.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWCvl9qKMfnurf9pg:-LEJWCvl9qKMfnurf9ph:b-2wcfy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/roachpb/internal.go#L64)):*
> ```Go
> }
> 
> // IsColumnar returns true if this InternalTimeSeriesData stores its samples
> ```

Copy pasta.

---

*[pkg/ts/db_test.go, line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWSwuA_s68II0cxGl:-LEJWSwuA_s68II0cxGm:b-k114hp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/db_test.go#L402)):*
> ```Go
> 		}
> 	}
> 	if a, e := len(actualDatapoints), expectedDatapointCount; a != e {
> ```

Did you move this because you didn't want the Fatal to hide possible errors from the data series test above?

---

*[pkg/ts/iterator_test.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWsx51lNPxofXe7wv:-LEJWsx51lNPxofXe7ww:b3bxld2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L97)):*
> ```Go
> 		StartTimestampNanos: startTimestamp,
> 		SampleDurationNanos: sampleDuration,
> 		Offset:              make([]int32, 0),
> ```

What are all of these zero size arrays for? You use append below, which happily accepts `nil`. Is this some weirdness where you want a zero-length array and not `nil` in the resulting TimeSeriesData?

---

*[pkg/ts/iterator_test.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXIcL3rExoYCsBQ78:-LEJXIcL3rExoYCsBQ79:b-u7t8jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L114)):*
> ```Go
> 	valuesForSample := make([]float64, 0, 1)
> 	for _, sample := range samples {
> 		offset := int32((sample.timestamp - startTimestamp) / sampleDuration)
> ```

OffsetForTimestamp?

---

*[pkg/ts/iterator_test.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXmBnDED8Ua3L6bqZ:-LEJXmBnDED8Ua3L6bq_:b48r9ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L130)):*
> ```Go
> 		} else {
> 			// Compute variance for previous sample if there was more than one
> 			// value.
> ```

Isn't there a one-off here where you never compute the compound quantities for the last sample (since there isn't a sample at the next offset that could trigger it)?

---

*[pkg/ts/iterator_test.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJYfeX1ZS0VS7i73Zl:-LEJYfeX1ZS0VS7i73Zm:b-5s4jxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L135)):*
> ```Go
> 				total := 0.0
> 				for i := range valuesForSample {
> 					total += math.Pow(valuesForSample[i]-avg, 2)
> ```

I assume you didn't want to bother with the running formulas for computing variance: https://www.johndcook.com/blog/standard_deviation/

(I think that's reasonable not to get into unless you expect huge aggregations here).

---

*[pkg/ts/iterator_test.go, line 496 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJZn99Bnwjm_O3FkO7:-LEJZn99Bnwjm_O3FkO8:b-4f5ipo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L496)):*
> ```Go
> 			samplePeriod: 10,
> 			downsampler:  tspb.TimeSeriesQueryAggregator_AVG,
> 			expectedDesc: []dataDesc{
> ```

These test cases are nicely compact, but they're way too dense for me to understand without looking up all sorts of numbers. Perhaps you can put a verbose comment on the first that explains the input and the output? The others can hopefully be transferred from that. (Will scroll past these changes for now).

---

*[pkg/ts/iterator_test.go, line 508 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJZiPoAtgkunl5RkDC:-LEJZiPoAtgkunl5RkDD:bo8xqu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L508)):*
> ```Go
> 			},
> 		},
> 		// Original sample period, max downsampler.  Should fill in max value.
> ```

Mega nit: double space.

---

*[pkg/ts/query.go, line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJ_Wrk9amPHSPMZjE1:-LEJ_Wrk9amPHSPMZjE2:b-8kl58z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L233)):*
> ```Go
> 
> // truncateSpan truncates the span underlying this iterator to the current iterator,
> // *not including* the current position.  That is, the logical underlying span
> ```

nitnit: double double space space.

---

*[pkg/ts/query.go, line 514 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJaHRp2S-NlEUsg6bv:-LEJaHRp2S-NlEUsg6bw:brtpnsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L514)):*
> ```Go
> 	if timespan.SampleDurationNanos != diskResolution.SampleDuration() {
> 		downsampleSpans(sourceSpans, timespan.SampleDurationNanos, query.GetDownsampler())
> 		// downsampleSpans always produces single-valued spans, so convert the downsampler to sum.
> ```

I'm not sure I understand this. Is the point here that with single-valued spans, all downsamples are the same? If so, why switch it to SUM? What's special about that one?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-:-LEJhgGcBwqWPtMHYwre:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/internal.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJW4Pp1S9al83y-fYj:-LELEExN1HveuDP6QhnW:b394i9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/roachpb/internal.go#L61)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Are we going to be storing all samples in columnar format when you're done, or only rollups?
</blockquote></details>

All samples.  For non-rollups, it will be a sparse columnar format (only Offset and Last columns).  However, there will still be some row-formatted data on disk.

---

*[pkg/roachpb/internal.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWCvl9qKMfnurf9pg:-LELJCBq-xHtaWcl19Lf:b-ippazx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/roachpb/internal.go#L64)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Copy pasta.
</blockquote></details>

Doh! Fixed these up.

---

*[pkg/ts/db_test.go, line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWSwuA_s68II0cxGl:-LELJXCE3Y9eb0RNNYRv:b-xz30yt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/db_test.go#L402)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Did you move this because you didn't want the Fatal to hide possible errors from the data series test above?
</blockquote></details>

Yes, more often than not when the expected data point didn't match, the model was actually correct.  What we want to catch with these two values are situations where both the model and the reality differ from the expectation - which likely indicates that you set your test up incorrectly, these are a sanity check.

This output pattern is better for debugging purposes.

---

*[pkg/ts/iterator_test.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJWsx51lNPxofXe7wv:-LELK35A3Wpve7evQgMT:bsltxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L97)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What are all of these zero size arrays for? You use append below, which happily accepts `nil`. Is this some weirdness where you want a zero-length array and not `nil` in the resulting TimeSeriesData?
</blockquote></details>

No reason, just missed this optimization. Removed the arrays.

---

*[pkg/ts/iterator_test.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXIcL3rExoYCsBQ78:-LELKQE4BNkO_epmpTh-:b-8ft9vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L114)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

OffsetForTimestamp?
</blockquote></details>

:100:

---

*[pkg/ts/iterator_test.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXmBnDED8Ua3L6bqZ:-LELOhwG6MRo3TArAFBJ:b-sff4kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L130)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Isn't there a one-off here where you never compute the compound quantities for the last sample (since there isn't a sample at the next offset that could trigger it)?
</blockquote></details>

Ah, good catch.  Refactored the variance computation into a closure, calling it both in the loop and once at the end.

---

*[pkg/ts/iterator_test.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJYfeX1ZS0VS7i73Zl:-LELKbfj6VAJK2n0-OZ9:b-rmoemx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L135)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I assume you didn't want to bother with the running formulas for computing variance: https://www.johndcook.com/blog/standard_deviation/

(I think that's reasonable not to get into unless you expect huge aggregations here).
</blockquote></details>

I was looking at exactly that, and for the real rollups I might use them if they are more efficient.  In this case, though, I decided that this is test code and it's much easier to read the explicit variance computation.

---

*[pkg/ts/iterator_test.go, line 496 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJZn99Bnwjm_O3FkO7:-LELSWYj-TsqBEw2PHzI:bo5hat4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L496)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

These test cases are nicely compact, but they're way too dense for me to understand without looking up all sorts of numbers. Perhaps you can put a verbose comment on the first that explains the input and the output? The others can hopefully be transferred from that. (Will scroll past these changes for now).
</blockquote></details>

Added a large comment explaining the structure of the test cases.  Reposted here for your convenience:

```go
// Each test case is structured as such:
	// + A description of an "input" span, which describes a list of
	// InternalTimeSeriesData structures that will be assembled into a data span.
	// Each structure has a start timestamp, a sample period (should be the same
	// for all structure), and a set of data samples.
	// + A sample period, which should be greater than or equal to the sample
	// period of the input span structures.
	// + A downsampler operation.
	// + A description of an "expected" span, which describes the list of
	// InternalTimeSeriesData structures that should result from running the input
	// span through the downsampling operation.
	//
	// Importantly, both the "input" and "expected" spans are defined using the
	// dataDesc structure, rather than explicitly creating InternalTimeSeriesData
	// structures. This is because we want to test both the row format and the
	// columnar format of InternalTimeSeriesData when downsampling - therefore,
	// using the descriptors, each test case is run using first row-structured
	// data, then column-structured data, and finally a mixed-format test which
	// combines the two. This gives us a very broad test area while still
	// maintaining a compact set of test cases.
```

---

*[pkg/ts/iterator_test.go, line 508 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJZiPoAtgkunl5RkDC:-LELSioVChP6jTChaj1_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L508)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Mega nit: double space.
</blockquote></details>

Done.

---

*[pkg/ts/query.go, line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJ_Wrk9amPHSPMZjE1:-LELQS6r9YzB3VMltyBs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L233)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nitnit: double double space space.
</blockquote></details>

Done.

---

*[pkg/ts/query.go, line 514 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJaHRp2S-NlEUsg6bv:-LELP4-P-aGn-tPMzIyD:b-ldxqz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L514)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm not sure I understand this. Is the point here that with single-valued spans, all downsamples are the same? If so, why switch it to SUM? What's special about that one?
</blockquote></details>

That's a good point, all of the current downsamplers are equivalent for single-valued samples.  That said, this won't always be the case: for example, a variance downsampler would always return 0 for a single-valued span.

Let me explain the though process behind this:
+ If we are actually downsampling data, then we pply the downsampler in the downsampleSpans function.
+ If, however, our requested resolution matches the on-disk resolution, then we don't need to call downsampleSpans  at all and can just pass directly to the aggregator. However, the aggregator then must be built to work with roll-up data - for example, if you request the max value, and your requested resolution matches the on-disk resolution of a rollup, the aggregator will use the "max" fields from the rollup.
+ So, following from those two points, if we perform the downsampling operation explicitly, we create a single-valued span (even if the on-disk data was a rollup) and then tell the aggregator to act as if the "sum" aggregator was being used - sum is arbitrarily chosen because it works trivially with single-valued spans.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26446)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ts/iterator_test.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXmBnDED8Ua3L6bqZ:-LEOtltF6FxIc296b8UR:bk2lfy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L130)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Ah, good catch.  Refactored the variance computation into a closure, calling it both in the loop and once at the end.
</blockquote></details>

And this didn't affect any test cases? This suggests that there's something that should be added.

---

*[pkg/ts/iterator_test.go, line 496 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJZn99Bnwjm_O3FkO7:-LEOu-v_097Za1Mg8yvP:b-wdj8zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L496)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Added a large comment explaining the structure of the test cases.  Reposted here for your convenience:

```go
// Each test case is structured as such:
	// + A description of an "input" span, which describes a list of
	// InternalTimeSeriesData structures that will be assembled into a data span.
	// Each structure has a start timestamp, a sample period (should be the same
	// for all structure), and a set of data samples.
	// + A sample period, which should be greater than or equal to the sample
	// period of the input span structures.
	// + A downsampler operation.
	// + A description of an "expected" span, which describes the list of
	// InternalTimeSeriesData structures that should result from running the input
	// span through the downsampling operation.
	//
	// Importantly, both the "input" and "expected" spans are defined using the
	// dataDesc structure, rather than explicitly creating InternalTimeSeriesData
	// structures. This is because we want to test both the row format and the
	// columnar format of InternalTimeSeriesData when downsampling - therefore,
	// using the descriptors, each test case is run using first row-structured
	// data, then column-structured data, and finally a mixed-format test which
	// combines the two. This gives us a very broad test area while still
	// maintaining a compact set of test cases.
```


</blockquote></details>

Nice, thanks!

---

*[pkg/ts/query.go, line 514 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJaHRp2S-NlEUsg6bv:-LEOuHb9BzI4Nakynid5:blo0xoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L514)):*
Got it, thanks! Though if you wanted to distill this explanation into a comment here, that would be appreciated. A line of the form

> // At the time of writing, all downsamplers are the identity on single-valued spans, but that may not be true forever (consider for instance a variance downsampler).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26446)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ts/iterator_test.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXmBnDED8Ua3L6bqZ:-LEQbRoj88PCd_mlqgzb:b-m0gf99) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L130)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

And this didn't affect any test cases? This suggests that there's something that should be added.
</blockquote></details>

Hmm, well we don't actually support the variance downsampler in the model yet (I pre-emptively added it here), and this algorithm here is actually test code.  I will be adding that support in my very next PR (which adds the ability to write columnar data), so this will get covered at that poin.

---

*[pkg/ts/query.go, line 514 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJaHRp2S-NlEUsg6bv:-LEQejaY-pacs_0RlX-U:bs7wumb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/query.go#L514)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Got it, thanks! Though if you wanted to distill this explanation into a comment here, that would be appreciated. A line of the form

> // At the time of writing, all downsamplers are the identity on single-valued spans, but that may not be true forever (consider for instance a variance downsampler).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26446)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/ts/iterator_test.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26446#-LEJXmBnDED8Ua3L6bqZ:-LEQg_lJ2ZIX8bC6yDbR:bk0ezuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8111e243c716ebbcdabedc35977db97370f200b/pkg/ts/iterator_test.go#L130)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Hmm, well we don't actually support the variance downsampler in the model yet (I pre-emptively added it here), and this algorithm here is actually test code.  I will be adding that support in my very next PR (which adds the ability to write columnar data), so this will get covered at that poin.
</blockquote></details>

Ah, makes sense.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26446)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
bors r+
